File: iosApp/iosApp/Presentation/Main/CafeManagement/CafeDashboard/CafeDashboardUiState.swift
//
//  CafeDashboardUiState.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/10.
//

import Foundation
import Shared

struct CafeDashboardUiState {
    var cafe: CafeDashboardData?
    var castPreviews: [CafeCastPreview] = []
    var pendingCastClaims: [PendingCastClaimPreview] = []
    var guestSchedules: [GuestCastSchedule] = []
    var externalLinks: [CafeDashboardExternalLink] = []
    var selectedCastId: String?
    var nextCastCursor: String?
    var hasMoreCasts = false
    var isLoadingMoreCasts = false
    var isDeleteCastDialogVisible = false
    var isExternalLinkSheetVisible = false
    var editingExternalLinkId: String?
    var externalLinkTitle = ""
    var externalLinkUrl = ""
    var instagramId = ""
    var twitterId = ""
    var tiktokId = ""
    var youtubeId = ""
    var isSavingSocialMedia = false
    var isSocialMediaSheetVisible = false
    var reservationUrl = ""
    var isReservationSheetVisible = false
    var isSavingReservation = false
    var isTableCountSheetVisible = false
    var isGuestSheetVisible = false
    var isGuestSaving = false
    var guestName = ""
    var guestProfileImage = ""
    var guestDate = ""
    var guestStartTime = CafeDashboardUiState.defaultGuestStartTime
    var guestEndTime = CafeDashboardUiState.defaultGuestEndTime
    var guestMemo = ""
    var guestDateOptions: [String] = []
    var guestTimeOptions: [String] = CafeDashboardUiState.defaultGuestTimeOptions()
    var currentTableCountInput = ""
    var totalTableCountInput = ""
    var isSavingTableCounts = false
    var isLoading = true
    var infoMessage: String?

    var isExternalLinkSubmitEnabled: Bool {
        !externalLinkTitle.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
        !externalLinkUrl.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    var isTableCountSubmitEnabled: Bool {
        guard let current = Int(currentTableCountInput), let total = Int(totalTableCountInput) else { return false }
        return total >= 0 && current >= 0 && current <= total
    }

    var isGuestSubmitEnabled: Bool {
        !guestName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
        !guestDate.isEmpty &&
        guestStartTime < guestEndTime
    }

    static let defaultGuestStartTime = "14:00"

    static let defaultGuestEndTime = "22:00"

    static func defaultGuestTimeOptions() -> [String] {
        (0...23).flatMap { hour in
            ["00", "30"].map { minute in String(format: "%02d:%@", hour, minute) }
        }
    }
}

struct CafeDashboardExternalLink: Identifiable {
    let id: String
    let title: String
    let url: String
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/cafemanagement/cafedashboard/CafeDashboardUiState.kt
package com.hhp227.concafe.presentation.main.cafemanagement.cafedashboard

import com.hhp227.concafe.domain.model.CafeCastPreview
import com.hhp227.concafe.domain.model.CafeDashboardData
import com.hhp227.concafe.domain.model.GuestCastSchedule
import com.hhp227.concafe.domain.model.PendingCastClaimPreview

data class CafeDashboardUiState(
    val cafe: CafeDashboardData? = null,
    val castPreviews: List<CafeCastPreview> = emptyList(),
    val pendingCastClaims: List<PendingCastClaimPreview> = emptyList(),
    val guestSchedules: List<GuestCastSchedule> = emptyList(),
    val externalLinks: List<CafeDashboardExternalLink> = emptyList(),
    val selectedCastId: String? = null,
    val nextCastCursor: String? = null,
    val hasMoreCasts: Boolean = false,
    val isLoadingMoreCasts: Boolean = false,
    val isDeleteCastDialogVisible: Boolean = false,
    val isExternalLinkSheetVisible: Boolean = false,
    val editingExternalLinkId: String? = null,
    val externalLinkTitle: String = "",
    val externalLinkUrl: String = "",
    val instagramId: String = "",
    val twitterId: String = "",
    val tiktokId: String = "",
    val youtubeId: String = "",
    val isSavingSocialMedia: Boolean = false,
    val isSocialMediaSheetVisible: Boolean = false,
    val reservationUrl: String = "",
    val isReservationSheetVisible: Boolean = false,
    val isSavingReservation: Boolean = false,
    val isTableCountSheetVisible: Boolean = false,
    val isGuestSheetVisible: Boolean = false,
    val isGuestSaving: Boolean = false,
    val guestName: String = "",
    val guestProfileImage: String = "",
    val guestDate: String = "",
    val guestStartTime: String = DEFAULT_GUEST_START_TIME,
    val guestEndTime: String = DEFAULT_GUEST_END_TIME,
    val guestMemo: String = "",
    val guestDateOptions: List<String> = emptyList(),
    val guestTimeOptions: List<String> = defaultGuestTimeOptions(),
    val currentTableCountInput: String = "",
    val totalTableCountInput: String = "",
    val isSavingTableCounts: Boolean = false,
    val isLoading: Boolean = true,
    val infoMessage: String? = null
) {
    val isExternalLinkSubmitEnabled: Boolean
        get() = externalLinkTitle.isNotBlank() && externalLinkUrl.isNotBlank()

    val isTableCountSubmitEnabled: Boolean
        get() {
            val current = currentTableCountInput.toIntOrNull() ?: return false
            val total = totalTableCountInput.toIntOrNull() ?: return false
            return total >= 0 && current >= 0 && current <= total
        }

    val isGuestSubmitEnabled: Boolean
        get() = guestName.isNotBlank() && guestDate.isNotBlank() && guestStartTime < guestEndTime

    companion object {
        const val DEFAULT_GUEST_START_TIME = "14:00"
        const val DEFAULT_GUEST_END_TIME = "22:00"
    }
}

data class CafeDashboardExternalLink(
    val id: String,
    val title: String,
    val url: String
)

private fun defaultGuestTimeOptions(): List<String> {
    return (0..23).flatMap { hour ->
        listOf("00", "30").map { minute -> hour.toString().padStart(2, '0') + ":$minute" }
    }
}
